File: app/models/project.rb
class Project < ApplicationRecord

  default_scope{ where(deleted: [false, nil])}


  validates :name, presence: true, uniqueness: true

  before_save :generate_key
  after_create :add_execution

  has_many :executions, dependent: :destroy
  has_many :testcases, dependent: :destroy
  has_many :environments, dependent: :destroy
  has_and_belongs_to_many :teams

  private

  def generate_key
    self.api_key = SecureRandom.hex
  end

  def add_execution
    execution = Execution.create(project_id: self.id, closed: false)
    return execution
  end
end

File: client/angular/src/app/domain/project.ts
import { TestCase } from "./testcases/testcase";
import { Keyword } from "./keyword";
import { Environment } from "./environment";
import { Execution } from "./executions/execution";


export class Project {

  constructor(
    public id: number,
    public name: string,
    public apiKey: string,
    public executionId: string,
    public testcases: Array<TestCase>,
    public executions: Array<Execution>,
    public keywords: Array<Keyword>,
    public environments: Array<Environment>
  ) {  }

  public static create(data: any): Project {
    let testcases = [];
    if (data.testcases) {
      for (let testcase of data.testcases) {
        testcases.push(TestCase.create(testcase));
      }
    }

    let environments = [];
    if (data.environments) {
      for (let environment of data.environments) {
        environments.push(Environment.create(environment));
      }
    }

    let executions = [];
    if (data.executions) {
      for (let execution of data.executions) {
        executions.push(Execution.create(execution));
      }
    }

    let keywords = [];
    if (data.keywords) {
      for (let keyword of data.keywords) {
        keywords.push(Keyword.create(keyword));
      }
    }

    return new Project(data.id, data.project_name, data.api_key, data.execution_id,
      testcases, executions, keywords, environments);
  }
}
